apps/CMakeLists.txt

```diff
@@ -1,30 +1,58 @@
-include_directories("${CMAKE_BINARY_DIR}/include")
-include_directories("${HALIDE_TOOLS_DIR}")
-
-if (MSVC)
-    # Sample apps, so we can disable some of the more pedantic warnings
-    add_definitions(-D_CRT_SECURE_NO_WARNINGS)
-    add_compile_options(/wd4244) # conversion from 'type1' to 'type2', possible loss of data.
-endif()
-
-add_subdirectory(autoscheduler)
-add_subdirectory(bilateral_grid)
-add_subdirectory(blur)
-add_subdirectory(c_backend)
-add_subdirectory(camera_pipe)
-add_subdirectory(conv_layer)
-add_subdirectory(glsl)
-add_subdirectory(harris)
-add_subdirectory(iir_blur)
-add_subdirectory(interpolate)
-add_subdirectory(lens_blur)
-add_subdirectory(linear_algebra)
-add_subdirectory(linear_blur)
-add_subdirectory(local_laplacian)
-add_subdirectory(nl_means)
-add_subdirectory(resize)
-add_subdirectory(stencil_chain)
-add_subdirectory(unsharp)
-
-# Don't add this one; it's deliberately standalone
-# add_subdirectory(wavelet)
+##
+# Test apps from the perspective of a consuming project.
+##
+
+# Set up a test fixture for installing the main project into a build directory
+
+set(HALIDE_DIR "${CMAKE_CURRENT_BINARY_DIR}/_Halide_install")
+
+# TODO: if we can support CMake 3.15+, we can use ${CMAKE_COMMAND} --install ...
```

Comment:
I am in favor of depending on the newest version, sure. 3.16 adds cool stuff like cross-platform precompiled headers.

Reply:
To second Benoit's complaint: Is 3.10 totally untenable? Requiring something not in the apt repo of the latest ubuntu LTS release is an extra burden for users.

Reply:
Well, the Makefile build is still around so we aren't requiring it.

Version 3.10 doesn't support the packaging features I'm using and intend to use. Object libraries (like those used to build the runtime) are broken before 3.14, as well. Kitware provides official support for up-to-date CMake on Ubuntu LTS and when admin access isn't available, pip can be used to install the most recent version locally.

Reply:
Also, there's a new Ubuntu LTS coming out in 8 weeks that will come with CMake 3.16.3.

Reply:
It also means we have to keep supporting and testing everything in the makefile build for the forseeable future. I'm not willing to require people to manually acquire newer versions of build tools before they can even get started - that's one yak too many.

Reply:
ok, so let's use 3.16.3 and just plan to continue to support and test the makefile build for a long time too.

Reply:
> I'm not willing to require people to manually acquire newer versions of build tools before they can even get started - that's one yak too many.

Get started _using_ Halide or _contributing_ to it? Those are two different scenarios.

Our python bindings build explodes if `imageio` isn't installed. And it's included by default, so the Makefile build fails if you haven't satisfied the requirements.txt tucked away in `python_bindings`. We could literally add `cmake>=3.14` to that file and developers would have to run the _same_ number of commands.

Reply:
Building Halide. Currently building a recent llvm is the main hurdle. It would be great if we could use an older llvm, but it's not feasible for us to support lots of llvm versions.

I don't consider the state of the python bindings to be a good target :)

Reply:
Ah, well, that story would change pretty substantially if we produced an APT or PIP package of Halide.

Reply:
For context on why I'm so conservative on build stuff: I recently had to build a Halide branch from scratch on a quad-core cortex-a53 arm dev board with ubuntu 16 to test something. Anything not in the ubuntu repo I have to build from source, and it takes approximately forever. I reimaged it as ubuntu 18 just to avoid building llvm, because Steven already had a prebuilt for arm ubuntu 18. It was then just a matter of building libHalide. It was still quite painful. And then of course my bug wouldn't repro. Halide, even as a jit compiler, is popular in these sorts of constrained environments, so we have to be very careful about requiring anything new.

Reply:
I'm not convinced that apt or pip would change life for the people currently building Halide. Aren't those the people who would use one of our releases?

Reply:
I wonder if cross compiling Halide would have been a better option in that case.